### PR TITLE
Add GoTestingDeclarationsInspection

### PR DESCRIPTION
**Pull request by eugene-auduchinok**
All testing functions (Test/Benchmark/Example) in test files should have appropriate signature (http://golang.org/pkg/testing/) and the first symbol after testing prefix must not be a lowercase letter.

I've created anonymous inner classes for quick fixes but I'm not sure if it had been better to create separated named classes.
Would it be better if I combine quick fixes into a single class? Or would it be better if I split inspection otherwise?

I haven't created tests yet as I'm not sure if I've done all these things right.

### CODE REVIEW

File: src/com/goide/inspections/GoTestingDeclarationsInspection.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+
+public class GoTestingDeclarationsInspection extends GoInspectionBase {
+  private enum TestingType {TEST, TEST_MAIN, BENCHMARK, EXAMPLE}
+
+  private static TestingType getTestingType(@NotNull String name) {
+    if (name.startsWith(GoConstants.TEST_PREFIX)) {
+      return name.toLowerCase().equals("testmain") ? TestingType.TEST_MAIN : TestingType.TEST;
```
Comment (zolotov) [inline src/com/goide/inspections/GoTestingDeclarationsInspection.java]: equalsIgnoreCase

File: src/com/goide/inspections/GoTestingDeclarationsInspection.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    }
+    if (name.startsWith(GoConstants.BENCHMARK_PREFIX)) return TestingType.BENCHMARK;
+    if (name.startsWith(GoConstants.EXAMPLE_PREFIX)) return TestingType.EXAMPLE;
+    return null;
+  }
+
+  private static String getPrefixFromType(@NotNull TestingType type) {
+    switch (type) {
```
Comment (zolotov) [inline src/com/goide/inspections/GoTestingDeclarationsInspection.java]: I believe TestingType can tell about its prefix without switch/case

File: src/com/goide/inspections/GoTestingDeclarationsInspection.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  @Override
+  protected void checkFile(@NotNull final GoFile file, @NotNull ProblemsHolder problemsHolder) {
+    if (!GoTestFinder.isTestFile(file)) return;
+    GoImportSpec testingImportSpec = file.getImportedPackagesMap().get(GoConstants.TESTING_PATH);
+    String testingAlias = null;
+    if (testingImportSpec != null) {
+      testingAlias = testingImportSpec.getAlias();
+      testingAlias = testingAlias == null ? testingImportSpec.getPath() : testingAlias;
```
Comment (zolotov) [inline src/com/goide/inspections/GoTestingDeclarationsInspection.java]: getLocalPackageName is more appropriate than getPath.
+ should be tested with dot-imports line `import . "testing"`

File: src/com/goide/inspections/GoTestingDeclarationsInspection.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+      final String prefix = getPrefixFromType(type);
+      final String testName = name.substring(prefix.length());
+      if (!StringUtil.isCapitalized(testName)) {
+        final PsiElement identifier = function.getIdentifier();
+        problemsHolder.registerProblem(identifier, "Test name must start with capital letter",
+                                       new LocalQuickFixBase("Capitalize " + testName) {
+                                         @Override
+                                         public void applyFix(@NotNull Project project, @NotNull ProblemDescriptor descriptor) {
+                                           identifier.replace(
```
Comment (zolotov) [inline src/com/goide/inspections/GoTestingDeclarationsInspection.java]: memory leak

**Comment by dlsniper** [comment]
I like the idea about checking for the capital letter.

However all the other parts of the signature are hard to catch because one could use

- gocheck
```
func (ds *DarkSide) TestSomething_Something(c *C) {}
```

- goconvey (maintains the standard test API)
```
func TestBowlingGameScoring(t *testing.T) {}
```

I'm not sure what other testing frameworks are doing but, aside from BDD ones, which all seem to have their own thing way different than the standard API.

**Comment by zolotov** [comment]
@dlsniper checking for capital letter is wrong approach. The right one implemented in GoTestFinder#checkPrefix

**Comment by zolotov** [comment]
> However all the other parts of the signature are hard to catch because one could use

This inspection works for gotest only

**Comment by eugene-auduchinok** [comment]
Oops. I didn't notice failing yesterday. Will fix it asap.

**Comment by zolotov** [comment]
Anyway looks good, tests are still needed though

File: src/com/goide/inspections/GoTestSignaturesInspection.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    GoImportSpec testingImportSpec = file.getImportedPackagesMap().get(GoConstants.TESTING_PATH);
+    String testingAlias = getTestingAlias(testingImportSpec);
+    for (GoFunctionDeclaration function : file.getFunctions()) {
+      GoTestFunctionType type = GoTestFunctionType.fromName(function.getName());
+      if (type == null) continue;
+      GoSignature signature = function.getSignature();
+      if (signature == null) continue;
+      List<GoParameterDeclaration> params = signature.getParameters().getParameterDeclarationList();
+      if (type == GoTestFunctionType.EXAMPLE) {
+        if (params.size() != 0) {
```
Comment (zolotov) [inline src/com/goide/inspections/GoTestSignaturesInspection.java]: !isEmpty

File: src/com/goide/inspections/GoTestSignaturesInspection.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        signature.replace(GoElementFactory.createFunctionSignatureFromText(project, ""));
+        return;
+      }
+      GoImportSpec testingImportSpec = (((GoFile)file).getImportedPackagesMap().get(GoConstants.TESTING_PATH));
+      if (testingImportSpec == null) {
+        ((GoFile)file).addImport(GoConstants.TESTING_PATH, null);
+      }
+      String paramType = getProperParamType(getTestingAlias(testingImportSpec), myType);
+      signature.replace(GoElementFactory.createFunctionSignatureFromText(project, myType.getParamType().toLowerCase() + " " + paramType));
+    }
```
Comment (zolotov) [inline src/com/goide/inspections/GoTestSignaturesInspection.java]: add locale to toLowerCase invokation

File: src/com/goide/inspections/GoTestSignaturesInspection.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+import com.intellij.util.containers.ContainerUtil;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+
+public class GoTestSignaturesInspection extends GoInspectionBase {
+  private static String getTestingAlias(GoImportSpec testingImportSpec) {
+    String alias = GoConstants.TESTING_PATH;
+    if (testingImportSpec != null) {
```
Comment (zolotov) [inline src/com/goide/inspections/GoTestSignaturesInspection.java]: alias variable is redundant

**Pull request merged**